packages/loopring_v3/contracts/aux/fast-withdrawal/FastWithdrawalLiquidityProvider.sol

```diff
@@ -81,21 +114,97 @@ contract FastWithdrawalLiquidityProvider is ReentrancyGuard, OwnerManagable
         if (token == address(0)) {
-            to.sendETHAndVerify(amount, gasleft()); // ETH
+            _balance = address(this).balance;
+            address(this).sendETHAndVerify(amount, gasleft()); // ETH
         } else {
-            token.safeTransferAndVerify(to, amount);  // ERC20 token
+            _balance = ERC20(token).balanceOf(address(this));
+            token.safeTransferAndVerify(address(this), amount);  // ERC20 token
         }
+        userTokenBalances[msg.sender][token] = userTokenBalances[msg.sender][token].add(amount);
+
+        uint _sharesToAdd = 0;
+        uint _totalShares = tokenShares[token];
+        if (_totalShares == 0) {
+            _sharesToAdd = 100 * SHARE_BASE;
+        } else {
+            _sharesToAdd = _totalShares.mul(amount) / _balance;
+        }
+
+        userTokenShares[msg.sender][token] = userTokenShares[msg.sender][token].add(_sharesToAdd);
+        tokenShares[token] = _balance.add(_sharesToAdd);
+        emit Deposit(msg.sender, token, amount);
+    }
+
+    // Allows the LP to transfer funds back out of this contract.
+    function requestWithdrawal(
+        address to,
+        address token,
+        uint    amount
+        )
+        external
+        nonReentrant
+    {
+        require(amount > 0, "ZERO_AMOUNT");
+        require(userTokenShares[msg.sender][token] > 0, "NOT_PARTICIPANT");
+        WithdrawalRequest memory _request = WithdrawalRequest(msg.sender, to, token, amount, block.timestamp);
+        bytes32 _hash = keccak256(
+            abi.encodePacked(
+                _request.from,
+                _request.to,
+                _request.token,
+                _request.amount,
+                _request.timestamp
+            )
+        );
+
+        require(withdrawalRequests[_hash].from == address(0), "DUPLICATED_REQUEST");
+        withdrawalRequests[_hash] = _request;
+        emit WithdrawalRequested(
+            _request.from,
+            _request.to,
+            _request.token,
+            _request.amount,
+            _request.timestamp
+        );
+
+    }
+
+    function processWithdrawal(bytes32 requestHash)
+        external
+        nonReentrant
+        onlyManager
+    {
+        processWithdrawalInternal(requestHash, false);
+    }
+
+    function forceWithdraw(bytes32 requestHash)
+        external
+        nonReentrant
+        onlyShutdown
+    {
+        processWithdrawalInternal(requestHash, true);
+    }
+
+    function notifyWithdrawalRequestTooOld(bytes32 requestHash)
+        external
+        onlyNotShutdown
+    {
+        WithdrawalRequest memory _request = withdrawalRequests[requestHash];
+        require(_request.timestamp > 0, "WITHDRAWAL_REQUEST_NOT_EXIST");
+        require(block.timestamp > _request.timestamp + MAX_AGE_WITHDRAWAL_REQUEST_UNTIL_SHUTDOWN,
+                "WITHDRAWAL_REQUEST_NOT_TOO_OLD");
+        shutdown = true;
+        emit LiguidityProviderShudown(address(this), block.timestamp);
     }
```

Comment:
Not sure how useful this is because we can always just never return the amount loaned. So no impact on trust with or without this force withdrawal feature. So depositors could simply be asking for a forced withdrawal on a contract with 0 funds.

Reply:
Yes, I agree. I will rethink the whole design.